presto-main/src/main/java/com/facebook/presto/execution/QueryLimitMemoryRevokingScheduler.java

```diff
@@ -0,0 +1,178 @@
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.function.Function;
+
+import static com.facebook.presto.execution.MemoryRevokingUtils.getMemoryPools;
+import static java.util.Objects.requireNonNull;
+
+public class QueryLimitMemoryRevokingScheduler
+{
+    private static final Logger log = Logger.get(QueryLimitMemoryRevokingScheduler.class);
+
+    private final ScheduledExecutorService taskManagementExecutor;
+    private final Function<QueryId, QueryContext> queryContextSupplier;
+
+    // Technically not thread safe but should be fine since we only call this on PostConstruct and PreDestroy.
+    // PreDestroy isn't called until server shuts down/ in between tests.
+    @Nullable
+    private ScheduledFuture<?> scheduledFuture;
+
+    private final AtomicBoolean checkPending = new AtomicBoolean();
+    private final List<MemoryPool> memoryPools;
+    private final MemoryPoolListener memoryPoolListener = this::onMemoryReserved;
+
+    @Inject
+    public QueryLimitMemoryRevokingScheduler(
+            LocalMemoryManager localMemoryManager,
+            SqlTaskManager sqlTaskManager,
+            TaskManagementExecutor taskManagementExecutor)
+    {
+        this(
+                ImmutableList.copyOf(getMemoryPools(localMemoryManager)),
+                requireNonNull(sqlTaskManager, "sqlTaskManager cannot be null")::getQueryContext,
+                requireNonNull(taskManagementExecutor, "taskManagementExecutor cannot be null").getExecutor());
+        log.debug("Using NodeLimitMemoryRevokingScheduler spilling strategy");
+    }
+
+    @VisibleForTesting
+    QueryLimitMemoryRevokingScheduler(
+            List<MemoryPool> memoryPools,
+            Function<QueryId, QueryContext> queryContextSupplier,
+            ScheduledExecutorService taskManagementExecutor)
+    {
+        this.memoryPools = ImmutableList.copyOf(requireNonNull(memoryPools, "memoryPools is null"));
+        this.queryContextSupplier = requireNonNull(queryContextSupplier, "queryContextSupplier is null");
+        this.taskManagementExecutor = requireNonNull(taskManagementExecutor, "taskManagementExecutor is null");
+    }
+
+    @PostConstruct
+    public void start()
+    {
+        registerPoolListeners();
+    }
+
+    @PreDestroy
+    public void stop()
+    {
+        if (scheduledFuture != null) {
+            scheduledFuture.cancel(true);
+            scheduledFuture = null;
+        }
+
+        memoryPools.forEach(memoryPool -> memoryPool.removeListener(memoryPoolListener));
+    }
+
+    @VisibleForTesting
+    void registerPoolListeners()
+    {
+        memoryPools.forEach(memoryPool -> memoryPool.addListener(memoryPoolListener));
+    }
+
+    private void onMemoryReserved(MemoryPool memoryPool, QueryId queryId)
+    {
+        try {
+            QueryContext queryContext = queryContextSupplier.apply(queryId);
+            long queryTotalMemoryUse = getTotalMemoryUse(queryContext);
```

Comment:
Seems like this is a bit unfortunate in that every time a query's memory limit increases it will contend for the `QueryContext` lock, the `MemoryPool` lock, release the `MemoryPool` lock, and then acquire the `MemoryPool` lock two more times inside of `MemoryPool#getTotalMemoryUse`.

Maybe `MemoryPool` listener should accept the memory reservation values as an argument so that it doesn't have to call right back into the `MemoryPool` to get them?